core/sql/ustat/hs_la.cpp

```diff
@@ -994,13 +999,118 @@ Lng32 HSHiveTableDef::DescribeColumnNames()
 }
 
 //=====================================================
+//
+NAString HSHbaseTableDef::getHistLoc(formatType format) const
+{
+  if ( HSGlobalsClass::isNativeHbaseCat(getCatName(format))) {
+    return HBASE_STATS_CATALOG "." HBASE_STATS_SCHEMA;
+  } else {  
+    NAString name(getCatName(format));
+    name.append(".");
+    name.append(getSchemaName(format));
+    return name;
+  }
+}
+
+static 
+Lng32 CreateExternalTable(const NAString& catName, const NAString& schName, const NAString& nativeTableName)
+{
+   HSLogMan *LM = HSLogMan::Instance();
+   if (LM->LogNeeded())
+      {
+        snprintf(LM->msg, sizeof(LM->msg), "Creating external table table for %s on demand.",
+                          nativeTableName.data());
+        LM->Log(LM->msg);
+      }
+
+   // do not have to worry about the catalog and schema for the new external table 
+   // here. These names will be determined by the processing logic. 
+   NAString ddl = "CREATE EXTERNAL TABLE ";
+   ddl.append(nativeTableName);
+   ddl.append(" FOR ");
+   ddl.append(catName);
+   ddl.append(".");
+   ddl.append(schName);
+   ddl.append(".");
+   ddl.append(nativeTableName);
+   ddl.append(" ATTRIBUTE ");
+
+   HSTranMan *TM = HSTranMan::Instance();
+   if (TM->InTransaction())
+      ddl.append(" AUDIT");
+   else
+      ddl.append("NO AUDIT");
+
+   Lng32 retcode = HSFuncExecDDL(ddl.data(), - UERR_INTERNAL_ERROR, NULL,
+                            "Create external table", NULL);
+
+   if (retcode < 0 && LM->LogNeeded())
+      {
+        snprintf(LM->msg, sizeof(LM->msg), "Creation of the external table failed.");
+        LM->Log(LM->msg);
+      }
+
+   return retcode;
+}
+
+NABoolean HSTableDef::setObjectUID(NABoolean createExternalTable)
+{
+  objectUID_ = naTbl_->objectUid().get_value();
+
+  if (createExternalTable && objectUID_ <= 0 && 
+      HSGlobalsClass::isNativeCat(getCatName(EXTERNAL_FORMAT)) ) {
+
+    // If objectUID is not set, it means there is no corresponding
+    // external table created for it. Need to create one here.
+    NAString catName = getCatName(EXTERNAL_FORMAT);
+    NAString schName = getSchemaName(EXTERNAL_FORMAT);
+    NAString objName = getObjectName(EXTERNAL_FORMAT);
+    Lng32 retcode = CreateExternalTable(catName, schName, objName);
+
+    if (retcode != 0)
+      return FALSE;
 
-NABoolean HSHbaseTableDef::objExists(labelDetail unused)
+    setNATable();
+    if (!naTbl_)
+      return FALSE;
+
```

Comment:
Do you get a reasonable error back if unable to create the external table or the special schemas?